### PR TITLE
Create code overview page in 2.0 manual

### PR DESCRIPTION
**Pull request by mikewalch**
This should help users/developers learn code and troubleshoot issues on their own.

Making this a draft PR to get input before I finish the page.

### CODE REVIEW

**Comment by milleruntime** [comment]
I like this idea!  We could also update this page, simplify it a bit: https://accumulo.apache.org/contributors-guide/

File: _docs-2/troubleshooting/code.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+The [Master] has the following responsibilities:
+
+  * [detect and respond][update] to [TabletServer] failures
+  * assign and balance tablets to Tablet Servers using a [TabletBalancer]
+  * handle table creation, alteration and deletion requests from clients using the [TableManager]
+  * coordinate changes to write-ahead logs using the [WalStateManager].
+  * report general status
+
+## Metadata Table
```
Comment (ctubbsii) [inline _docs-2/troubleshooting/code.md]: Accumulo has two metadata tables. This section should indicate that `accumulo.metadata` (id: `!0`) contains metadata for user tables, and that `accumulo.root` (id: `+r`) contains metadata for `accumulo.metadata`, and that `ZooKeeper` is used to store metadata for the `accumulo.root` table.

File: _docs-2/troubleshooting/code.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+## Garbage Collector
+
+The [GarbageCollector] has the following repsonsibilities:
+
+  * [identify RFiles] in HDFS that are no longer needed and delete them
+  * multiple garbage collectors can be run to provide hot-standby support
+
+[Master]: {% ghcu server/master/src/main/java/org/apache/accumulo/master/Master.java %}
+[update]: {% ghcu server/master/src/main/java/org/apache/accumulo/master/Master.java#L1332 %}
+[retrieves]: {% ghcu server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java#L1251 %}
```
Comment (ctubbsii) [inline _docs-2/troubleshooting/code.md]: Pointing to specific line numbers is going to bite us as edits are made. Are these pointing to a tag or other fixed blob?

Comment (mikewalch) [reply _docs-2/troubleshooting/code.md]: Right now, these are pointing to master branch but when 2.0 is released, I will have them point to 2.0 branch.

Comment (mikewalch) [reply _docs-2/troubleshooting/code.md]: The branch is [set in _config.yml](https://github.com/apache/accumulo-website/blob/master/_config.yml#L75) so it's easy to update.  You can also hard code a specific branch (see the REAMDE.md).

File: _docs-2/troubleshooting/code.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+  * [receives writes] from clients
+  * [retrieves] the [Tablet] that should be written to
+  * [persists writes] to a write-ahead log using [TabletServerLogger]
+  * sorts new key/value pairs in memory
+  * periodically [flush] sorted key/value pairs to new RFiles in HDFS
+  * responds to reads from clients and form a sorted merge view of all
+    key/value pairs from all files & memory
+  * perform recovery of a [Tablet] that was previously on a server that failed
+    and reapply any writes found in the write-ahead log to the tablet
```
Comment (keith-turner) [inline _docs-2/troubleshooting/code.md]: Should add compactions to the list.  Could also add caching and bulk import.

File: _docs-2/troubleshooting/code.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+## Garbage Collector
+
+The [GarbageCollector] has the following repsonsibilities:
+
+  * [identify RFiles] in HDFS that are no longer needed and delete them
+  * multiple garbage collectors can be run to provide hot-standby support
```
Comment (keith-turner) [inline _docs-2/troubleshooting/code.md]: Also GCs WALs

File: _docs-2/troubleshooting/code.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+started.
+
+## Master
+
+The [Master] has the following responsibilities:
+
+  * [detect and respond][update] to [TabletServer] failures
+  * assign and balance tablets to Tablet Servers using a [TabletBalancer]
+  * handle table creation, alteration and deletion requests from clients using the [TableManager]
+  * coordinate changes to write-ahead logs using the [WalStateManager].
```
Comment (keith-turner) [inline _docs-2/troubleshooting/code.md]: Could be more specific.  It coordinates sorting of WALs and associates WALs to tablets that need them (tablets that were previously assigned to a dead tserver) before assigning.

**Comment by keith-turner** [comment]
>The more code we add in place of static website content, the more the bar is raised, making it more difficult for others (especially non-programmers) to contribute to the documentation.

It may raise the bar for contributing to the website, but I think its worthwhile if it results in a more productive experience for users.  W/o the new automation I would not create many links that I now create because it would be too cumbersome.  I think these links are beneficial to the user.

File: _docs-2/troubleshooting/code.md (hunk 2 above)
Comment (ctubbsii) [reply _docs-2/troubleshooting/code.md]: Okay, sounds good. As long as the line numbers are accurate when it's tagged.

**Comment by milleruntime** [comment]
@mikewalch any plans to complete this?  If not, I will just close this for now since you were championing for it.  We can always open another PR in the future if you pick it back up.

**Pull request closed without merging**